### PR TITLE
Add Redis Data Cacher

### PR DESCRIPTION
**Pull request by howhelan**

### CODE REVIEW

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **howhelan**
You can assign the PR to them by writing `/assign @howhelan` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=pauwels-labs%2Fredact-data).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/pauwels-labs/redact-data/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: src/cache/redis.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+use std::time::Duration;
+use mobc_redis::{redis, RedisConnectionManager};
+use mobc::{Connection, Pool};
+use mobc_redis::redis::{AsyncCommands, ToRedisArgs, FromRedisValue, RedisWrite, RedisResult, Value, from_redis_value, ErrorKind};
+use crate::Data;
+
+pub type MobcPool = Pool<RedisConnectionManager>;
+pub type MobcCon = Connection<RedisConnectionManager>;
+
+const CACHE_POOL_MAX_OPEN: u64 = 16;
```
Comment (ajpauwels) [inline src/cache/redis.rs]: make config struct, pass it in constructor, load the config struct in main from config file

File: src/storage.rs [hunk 2]
```diff
@@ -41,17 +43,72 @@ where
...
+        }
+    }
+}
+
+#[async_trait]
+impl<T: DataStorer, V: DataCacher> DataStorer for CachedDataStorer<T, V> {
+    async fn get(&self, path: &str) -> Result<Data, DataStorerError> {
+        let cache_hit = self.cacher.exists(path).await?;
+        if cache_hit {
+            self.cacher.expire(path, 60).await?;
```
Comment (ajpauwels) [inline src/storage.rs]: turn 60 into config value

File: src/storage.rs [hunk 3]
```diff
@@ -41,17 +43,72 @@ where
...
+            cacher,
+        }
+    }
+}
+
+#[async_trait]
+impl<T: DataStorer, V: DataCacher> DataStorer for CachedDataStorer<T, V> {
+    async fn get(&self, path: &str) -> Result<Data, DataStorerError> {
+        let cache_hit = self.cacher.exists(path).await?;
+        if cache_hit {
```
Comment (ajpauwels) [inline src/storage.rs:80]: double check if redis will return a not found if the data has expired

**Review by ajpauwels — CHANGES_REQUESTED** [review]

File: src/storage.rs (hunk 3 above)
Comment (howhelan) [reply src/storage.rs:80]: per redis EXPIRE command docs:

> Set a timeout on key. After the timeout has expired, the key will automatically be deleted.

File: src/storage.rs (hunk 2 above)
Comment (howhelan) [reply src/storage.rs]: done

File: src/cache/redis.rs (hunk 1 above)
Comment (howhelan) [reply src/cache/redis.rs]: done

**Review by ajpauwels — APPROVED** [review]

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[ajpauwels](https://github.com/pauwels-labs/redact-data/pull/2#pullrequestreview-682314283 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=pauwels-labs%2Fredact-data).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/pauwels-labs/redact-data/blob/main/OWNERS)~~ [ajpauwels]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by capt-haddock** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[ajpauwels](https://github.com/pauwels-labs/redact-data/pull/2#issuecomment-860032157 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=pauwels-labs%2Fredact-data).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/pauwels-labs/redact-data/blob/main/OWNERS)~~ [ajpauwels]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**